diff --git a/board/ti/am64x/am64x.env b/board/ti/am64x/am64x.env
index 53c7d8d6790..806be23559b 100644
--- a/board/ti/am64x/am64x.env
+++ b/board/ti/am64x/am64x.env
@@ -18,7 +18,7 @@ findfdt=
 		echo WARNING: Could not determine device tree to use; fi;
 	setenv fdtfile ${name_fdt}
 name_kern=Image
-console=ttyS2,115200n8
+console=ttyS0,115200n8
 args_all=setenv optargs earlycon=ns16550a,mmio32,0x02800000 ${mtdparts}
 run_kern=booti ${loadaddr} ${rd_spec} ${fdtaddr}
 
